compiler/src/dotty/tools/dotc/transform/ShortcutImplicits.scala

```diff
@@ -0,0 +1,143 @@
+package dotty.tools.dotc
+package transform
+
+import TreeTransforms._
+import core.DenotTransformers.IdentityDenotTransformer
+import core.Symbols._
+import core.Contexts._
+import core.Types._
+import core.Flags._
+import core.Decorators._
+import core.StdNames.nme
+import core.Names._
+import core.NameOps._
+import ast.Trees._
+import ast.tpd
+import collection.mutable
+
+/** This phase optimizes code using implicit function types, by applying two rewrite rules.
+ *  Let IF be the implicit function type
+ *
+ *      implicit Us => R
+ *
+ *  (1) A method definition
+ *
+ *      def m(xs: Ts): IF = implicit (ys: Us) => E
+ *
+ *  is expanded to two methods:
+ *
+ *      def m(xs: Ts): IF = implicit (ys: Us) => m$direct(xs)(ys)
+ *      def m$direct(xs: Ts)(ys: Us): R = E
```

Comment:
Let's say you have

     val cl: implicit A => B
     def f: implicit A => B = cl

Then by the rules of implicit expansion this will gve you an eta-expansion on the rhs of `f`:

    def f: implicit A => B = implicit $x: A => cl($x)

That's what guarantees that the right-hand side of every method with implicit function result type is an implicit function value. For normal functions that guarantee does not hold,
so we might well be pessimizing code with a shortcutting optimization.